### ISSUE TITLE
Long division is extremly slow

### ISSUE THREAD

**Issue by terminal-cs** [CosmosOS/Cosmos#2355]
#### Area of Cosmos - What area of Cosmos are we dealing with?
Division, (implemented in x#)

#### Expected Behaviour - What do you think that should happen?
numbers should divide fast without much performance lost

#### Actual Behaviour - What unexpectedly happens?
significant speed impact (from 600 fps all the way down to 40)

#### Reproduction - How did you get this error to appear?
34671 / 1024 every frame

#### Version - Were you using the User Kit or Dev Kit? And what User Kit version or Dev Kit commit (Cosmos, IL2CPU, X#)?
devkit

**Comment by zarlo** [comment]
it looks like we are using the div opcode from nasm i dont know if we can make it faster in a way that makes sense

**Comment by quajak** [comment]
The issue is with long not int, where we cannot use the div assembly instruction.

**Comment by AnErrupTion** [comment]
Is this really a bug? Division works, it's just slow, so I don't think it can be considered a bug. @terminal-cs @quajak

**Comment by quajak** [comment]
Well if its too slow to be practical then I would consider it a bug

**Comment by AnErrupTion** [comment]
I'm not able to reproduce this bug. I'm even using C# division and have set fields to ``long``, yet it did not impact performance. So, I'm assuming that using X# directly will not have any impact either. Could you give a bit more details please? @terminal-cs

**Comment by terminal-cs** [comment]
> I'm not able to reproduce this bug. I'm even using C# division and have set fields to `long`, yet it did not impact performance. So, I'm assuming that using X# directly will not have any impact either. Could you give a bit more details please? @terminal-cs

I don't know what else i need to say here, all i know is that i divided a big number every frame and it had a huge performance impact

**Comment by AnErrupTion** [comment]
> > I'm not able to reproduce this bug. I'm even using C# division and have set fields to `long`, yet it did not impact performance. So, I'm assuming that using X# directly will not have any impact either. Could you give a bit more details please? @terminal-cs
> 
> I don't know what else i need to say here, all i know is that i divided a big number every frame and it had a huge performance impact

Is the division that you gave (``34671 / 1024``) the actual one you did every frame or is it another one?

**Comment by terminal-cs** [comment]
> > > I'm not able to reproduce this bug. I'm even using C# division and have set fields to `long`, yet it did not impact performance. So, I'm assuming that using X# directly will not have any impact either. Could you give a bit more details please? @terminal-cs
> > 
> > 
> > I don't know what else i need to say here, all i know is that i divided a big number every frame and it had a huge performance impact
> 
> Is the division that you gave (`34671 / 1024`) the actual one you did every frame or is it another one?

another one I belive, I was dividing some memory variable so get it in mb

**Comment by AnErrupTion** [comment]
> > > > I'm not able to reproduce this bug. I'm even using C# division and have set fields to `long`, yet it did not impact performance. So, I'm assuming that using X# directly will not have any impact either. Could you give a bit more details please? @terminal-cs
> > > 
> > > 
> > > I don't know what else i need to say here, all i know is that i divided a big number every frame and it had a huge performance impact
> > 
> > 
> > Is the division that you gave (`34671 / 1024`) the actual one you did every frame or is it another one?
> 
> another one I belive, I was dividing some memory variable so get it in mb

Ah, so you were getting the amount of memory... I'll try that soon.

**Comment by AnErrupTion** [comment]
I tried with ``Result = GCImplementation.GetUsedRAM() / 1024;`` and yet my performance is practically the same (no major drop in performance).

**Comment by quajak** [comment]
GetUsedRAM returns an uint not a long

**Comment by AnErrupTion** [comment]
> GetUsedRAM returns an uint not a long

Yeah but isn't it the way to get used RAM? It's also what OP used (at least, getting used RAM) for their division.

**Comment by ascpixi** [comment]
Practically, 64-bit division while running in protected mode will always be slower than 32-bit division - just like 32-bit division is slower while in real mode. The only practical solution would be to use long mode instead. Until then, we can take a look at the GCC function `_divdi3` and try to implement it in IL2CPU.

**Comment by valentinbreiz** [comment]
Cosmos Gen3 has replaced the IL2CPU-based Gen2 codebase, and `gen3` is now the default branch of this repository. This issue was filed against Gen2, which is no longer developed, so it is being closed as part of the Gen3 cutover (#3088).

If you still hit this on Gen3, please open a fresh issue. The compiler, the build pipeline and the debugging story are all different now, so a new report will get a far more useful answer than reviving this one. Start here: https://cosmosos.github.io/

Gen2 remains on the `gen2` branch, and this issue stays readable.
